Refactor Faiss-based vector format for easier backport

kaivalnp opened the pull request:
### Description

Refactor classes of the Faiss-based vector format to simplify backport to 10.x
- Extract minimal functionality required for the format into a new `FaissLibrary` interface
- Separate all function calls to the native library into a new `FaissNativeWrapper` (and use `invokeExact` for faster calls!)
- Create a new`FaissLibraryNativeImpl` class implementing `FaissLibrary` with `FaissNativeWrapper` under the hood
- Dynamically load `FaissLibraryNativeImpl` from `FaissLibrary` at runtime
- This setup encapsulates "unsafeness" into `FaissNativeWrapper` and `FaissLibraryNativeImpl` (as marked by TODOs in #14907), which can _independently_ be moved into the `java21/` directory for an easier backport (see comments in #14843), while the `FaissKnnVectors{Format/Reader/Writer}` can stay as-is!
- Improve some error handling

kaivalnp committed:
Refactor Faiss-based vector format for easier backport

kaivalnp committed:
Fix errorprone

mikemccand commented on lucene/sandbox/src/java/org/apache/lucene/sandbox/codecs/faiss/FaissLibrary.java:
```diff
@@ -0,0 +1,83 @@
+interface FaissLibrary {
+  FaissLibrary INSTANCE = lookup();
+
+  // TODO: Use vectorized version where available
```
Hmm what does this `TODO` mean again?  Is "vectorized" meaning "SIMD instructions"?  This term (vector) is overloaded!   https://youtu.be/fVq4_HhBK8Y

Maybe clarify to `// TODO: use SIMD Faiss API versions where available` or so (if that's what it really means)?  I think this is necessary because the Faiss build process will produce specific dynamic library for specific SIMD targets (AVX-512 vs AVX-128 etc.)?

mikemccand reviewed (approved):
Thank you @kaivalnp -- this looks clean to me -- but @uschindler probably has suggestions still ;)

uschindler reviewed (approved):
To me the whole thing looks fine. I have not checked the internals and only looked at the native code. It looks fine to have everything which calls native code into two classes.

The try/catch code looks correct. The redundancy here is needed, thanks for adding it, using a wrapper method for catching exceptions as before was not a good idea, because it required to use slow/inexact method handles and varargs.

But please remove the lookup of the native implementation in FaiddLibrary.java. In Lucene's main branch this is useless. Just return the instance directly (see code of MMapDir in main branch, there are no such indirections anymore).

Like Robert, I don't agree with backporting this to 10.x branch. This should be a new feature for Lucene 11.

kaivalnp committed:
Address comments

- Avoid dynamic lookup for FaissLibraryNativeImpl on the main branch
- Explain the SIMD situation in more detail

kaivalnp committed:
Merge branch 'main' into refactor-faiss

kaivalnp commented on lucene/sandbox/src/java/org/apache/lucene/sandbox/codecs/faiss/FaissLibrary.java:
```diff
@@ -0,0 +1,83 @@
+  String NAME = "faiss_c";
+  String VERSION = "1.11.0";
+
+  private static FaissLibrary lookup() {
```
Makes sense, updated

kaivalnp commented:
Thanks a lot @uschindler!

> I don't agree with backporting this to 10.x branch

Personally, I feel that the main blocker here was some differences in function names that allocated / read strings and arrays from native memory (which I think could be worked around using method handles to point to the correct function) -- but I agree, it could become messy due to the incompatibility.

I'll hold off on trying to backport this change unless someone feels otherwise.

mikemccand reviewed (approved):
Thanks @kaivalnp -- this looks cleaner!

It sounds like consensus is to not backport this to 10.x, and rather release 11.0 soon (end of year ish?).

uschindler reviewed (approved):
Latest changes look OK.

mikemccand commented:
Thank you @kaivalnp!
